### PR TITLE
fix connected check if no firewall hook exists

### PR DESCRIPTION
**Pull request by mafintosh**
I'll add a test as well, so don't merge yet

### CODE REVIEW

**Comment by mafintosh** [comment]
wait, this actually might be fine as is, i think i misunderstood the code haha

File: src/udx.c [hunk 1]
```diff
@@ -644,8 +644,10 @@ process_packet (udx_socket_t *socket, char *buf, ssize_t buf_len, struct sockadd
 
   // We expect this to be a stream packet from now on
 
-  if (!(stream->status & UDX_STREAM_CONNECTED) && (stream->on_firewall != NULL && stream->on_firewall(stream, socket, addr))) {
-    return 1;
+  if (!(stream->status & UDX_STREAM_CONNECTED)) {
+    if (stream->on_firewall != NULL && stream->on_firewall(stream, socket, addr)) return 1;
```
Comment (kasperisager) [inline src/udx.c:648]:
```suggestion
    if (stream->on_firewall == NULL || stream->on_firewall(stream, socket, addr)) return 1;
```

Early bail if the firewall hook isn't set.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: src/udx.c (hunk 1 above)
Comment (mafintosh) [reply src/udx.c:648]: right!

**Comment by mafintosh** [comment]
@kasperisager actually think this is fine as it was originally now, before my PR, cause we can buffer incoming packets without being connected, and we guard when responding with the state packet whether are connected or not. gonna think about it a bit tho.

**Comment by kasperisager** [comment]
The test case that now fails (https://github.com/hyperswarm/libudx/blob/79435797c2ca36256dbb754e1aebb16097ba425b/test/stream.js#L533-L564) actually explicitly relies on the current behaviour. That is, the firewall hook returns `false`, but doesn't connect, to indicate that incoming control packets should be processed.

Come to think of it, the return value of the firewall hook doesn't really make sense if the only signal that's actually considered is whether the stream connected or not. That's of course also a perfectly fine design; if the stream is connected after the firewall hook has run then proceed, otherwise abort. It does mean that it won't be possible to accept just incoming control packets, buffering data packets and outgoing control packets for when the stream later connects.

**Comment by mafintosh** [comment]
Agreed, closing

**Pull request closed without merging**